tests/amqpprox_session.t.cpp

```diff
@@ -1502,6 +1521,85 @@ TEST_F(SessionTest,
     EXPECT_TRUE(session->finished());
 }
 
+TEST_F(SessionTest, Forward_Received_Close_Method_To_Client_During_Handshake)
+{
+    EXPECT_CALL(d_selector, acquireConnection(_, _))
+        .WillOnce(DoAll(SetArgPointee<0>(d_cm), Return(0)));
+
+    EXPECT_CALL(*d_mapper, prime(_, _)).Times(AtLeast(1));
+    EXPECT_CALL(*d_mapper, mapToHostname(makeEndpoint("2.3.4.5", 2345)))
+        .WillRepeatedly(Return(std::string("host1")));
+    EXPECT_CALL(*d_mapper, mapToHostname(makeEndpoint("1.2.3.4", 1234)))
+        .WillRepeatedly(Return(std::string("host0")));
+    EXPECT_CALL(*d_mapper, mapToHostname(makeEndpoint("1.2.3.4", 32000)))
+        .WillRepeatedly(Return(std::string("host0")));
+    EXPECT_CALL(*d_mapper, mapToHostname(makeEndpoint("3.4.5.6", 5672)))
+        .WillRepeatedly(Return(std::string("host2")));
+
+    TestSocketState::State base;
+    base.d_local  = makeEndpoint("1.2.3.4", 1234);
+    base.d_remote = makeEndpoint("2.3.4.5", 2345);
+    base.d_secure = false;
+
+    TestSocketState::State clientBase;
+    clientBase.d_local  = makeEndpoint("1.2.3.4", 32000);
+    clientBase.d_remote = makeEndpoint("3.4.5.6", 5672);
+    clientBase.d_secure = false;
+
+    // Initialise the state
+    d_serverState.pushItem(0, base);
+    driveTo(0);
+
+    testSetupServerHandshake(1);
+
+    // Read a protocol header from the client and reply with Start method
+    // Client  ----AMQP Header--->  Proxy                         Broker
+    // Client  <-----Start--------  Proxy                         Broker
+    testSetupClientSendsProtocolHeader(2);
+
+    // Client  ------StartOk----->  Proxy                         Broker
+    // Client  <-----Tune---------  Proxy                         Broker
+    testSetupClientStartOk(3);
+
+    // Client  ------TuneOk------>  Proxy                         Broker
+    // Client  ------Open-------->  Proxy                         Broker
+    testSetupClientOpen(4);
+
+    // Client                       Proxy  <----TCP CONNECT---->  Broker
+    testSetupProxyConnect(5, &clientBase);
+
+    // Client                       Proxy  <-----HANDSHAKE----->  Broker
+    // Client                       Proxy  -----AMQP Header---->  Broker
+    testSetupProxySendsProtocolHeader(6);
+
+    // Client                       Proxy  <-------Start--------  Broker
+    // Client                       Proxy  --------StartOk----->  Broker
+    testSetupProxySendsStartOk(7, "host1", 2345, LOCAL_HOSTNAME, 1234, 32000);
+
+    // Client  <-------Close------- Proxy  <-------Close--------  Broker
+    testSetupProxyReceiveClose(8);
+
```

Comment:
So basically this function simulates the method exchanges, when amqpprox receives the close method from server. So basically, first amqpprox starts handshake with client. And when amqpprox receives Open method, amqpprox initiate handshake with actual server and start sending same received frames to server.

When amqpprox sends the StartOk method, amqpprox receives Close method from the server (Inside `testSetupProxyReceiveClose ` function `d_clientState.pushItem` call) (here one can assume that amqpprox got the Close method because of invalid credentials specified in StartOk method)

So now amqpprox will sends the Close method to client and then shutdown and close sockets for both clients and server. Specified all expectations inside `testSetupProxyReceiveClose ` function `d_serverState.expect`  and `d_clientState.expect` call.

Here I am testing with default Close method. But I can modify test to use custom Close method with dummy reason and error code to test more thoroughly.